File: solver.py
import numpy as np


def solve(sudoku: np.ndarray) -> bool:
    """
    Solves the given sudoku in-place using recursive backtracking.
    Returns True if a solution was found, False otherwise.
    """

    row_contains: list[set[int]] = [set() for _ in range(9)]
    col_contains: list[set[int]] = [set() for _ in range(9)]
    box_contains: list[set[int]] = [set() for _ in range(9)]

    # Helper function
    # Returns a tuple with the row column and box
    def getPosition(index: int) -> tuple[int, int, int]:
        row = index // 9
        col = index % 9
        box = (row // 3) * 3 + col // 3

        return (row, col, box)

    # Add existing numbers
    for i in range(81):
        row, col, box = getPosition(i)
        num: int = sudoku[row, col]

        if num == 0:
            continue

        row, col, box = getPosition(i)
        row_contains[row].add(num)
        col_contains[col].add(num)
        box_contains[box].add(num)

    # Recursive solve function
    def _solve(index: int) -> bool:

        # Loop until next empty cell
        row, col, box = getPosition(index)
        while sudoku[row, col] != 0:
            index += 1
            row, col, box = getPosition(index)
            if index >= 81:
                return True

        # Get invalid numbers for cell
        invalid_numbers: set[int] = row_contains[row] \
            | col_contains[col] \
            | box_contains[box]

        # If no number is available, backtrack
        if len(invalid_numbers) == 9:
            return False

        # Try all numbers
        for num in range(1, 10):
            if num in invalid_numbers:
                continue

            sudoku[row, col] = num
            row_contains[row].add(num)
            col_contains[col].add(num)
            box_contains[box].add(num)

            if (_solve(index)):
                return True

            row_contains[row].discard(num)
            col_contains[col].discard(num)
            box_contains[box].discard(num)

        sudoku[row, col] = 0
        return False

    return _solve(0)

File: website/app/util/solver.ts
/**
 * Solves the given sudoku in-place using recursive backtracking.
 * Returns True if a solution was found, False otherwise.
 */
export function solve(sudoku: Uint8Array): boolean {
    console.log("solving")
    const row_contains: Set<number>[] = Array(9).fill(0).map(() => new Set())
    const col_contains: Set<number>[] = Array(9).fill(0).map(() => new Set())
    const box_contains: Set<number>[] = Array(9).fill(0).map(() => new Set())

    /**
     * Helper function
     * Returns a tuple with the row column and box
     */
    function getPosition(index: number): number[] {
        const row = ~~(index / 9)
        const col = index % 9
        const box = ~~(row / 3) * 3 + ~~(col / 3)

        return [row, col, box]
    }

    // Add existing numbers
    for (let i = 0; i < 81; i++) {
        const [row, col, box] = getPosition(i)
        const num = sudoku[i]

        if (num == 0)
            continue;

        row_contains[row].add(num)
        col_contains[col].add(num)
        box_contains[box].add(num)
    }

    /**
     * Recursive solve function
     */
    function _solve(index: number): boolean {

        // Loop until next empty cell
        let [row, col, box] = getPosition(index)
        while (sudoku[index] != 0) {
            index++;
            [row, col, box] = getPosition(index)
            if (index >= 81)
                return true
        }

        // Get invalid numbers for cell
        const invalid_numbers: Set<number> = new Set([...Array.from(row_contains[row]), ...Array.from(col_contains[col]), ...Array.from(box_contains[box])])

        // If no number is avaliable, backtrack
        if (invalid_numbers.size == 9)
            return false

        // Try all numbers
        for (let num = 1; num < 10; num++) {
            if (invalid_numbers.has(num))
                continue

            sudoku[index] = num
            row_contains[row].add(num)
            col_contains[col].add(num)
            box_contains[box].add(num)

            if (_solve(index))
                return true

            row_contains[row].delete(num)
            col_contains[col].delete(num)
            box_contains[box].delete(num)
        }
        sudoku[index] = 0
        return false
    }

    // Call recursive function 
    return _solve(0)
}
